cdap-common/src/main/java/io/cdap/cdap/common/internal/remote/RemoteClientFactory.java

```diff
@@ -17,51 +17,98 @@
+    this(discoveryClient, internalAuthenticator, new NoOpRemoteAuthenticator(),
+        "");
   }
 
   @Inject
+  public RemoteClientFactory(DiscoveryServiceClient discoveryClient,
+      InternalAuthenticator internalAuthenticator,
+      RemoteAuthenticator remoteAuthenticator, CConfiguration cConf) {
+    this(discoveryClient, internalAuthenticator, remoteAuthenticator, "",
+        cConf.getBoolean(InternalRouter.USE_INTERNAL_ROUTER));
+  }
+
   public RemoteClientFactory(DiscoveryServiceClient discoveryClient,
       InternalAuthenticator internalAuthenticator,
       RemoteAuthenticator remoteAuthenticator) {
-    this(discoveryClient, internalAuthenticator, remoteAuthenticator, "");
+    this(discoveryClient, internalAuthenticator, remoteAuthenticator, "",
+        false);
   }
 
+
   public RemoteClientFactory(DiscoveryServiceClient discoveryClient,
       InternalAuthenticator internalAuthenticator,
       RemoteAuthenticator remoteAuthenticator, String pathPrefix) {
+    this(discoveryClient, internalAuthenticator, remoteAuthenticator,
+        pathPrefix, false);
+  }
+
+  public RemoteClientFactory(DiscoveryServiceClient discoveryClient,
+      InternalAuthenticator internalAuthenticator,
+      RemoteAuthenticator remoteAuthenticator, String pathPrefix,
+      boolean useInternalRouter) {
     this.discoveryClient = discoveryClient;
     this.internalAuthenticator = internalAuthenticator;
     this.remoteAuthenticator = remoteAuthenticator;
     this.pathPrefix = pathPrefix;
+    this.useInternalRouter = useInternalRouter;
   }
 
   public RemoteClient createRemoteClient(String discoverableServiceName,
-      HttpRequestConfig httpRequestConfig,
-      String basePath) {
-    basePath = basePath.startsWith("/") ? pathPrefix + basePath : pathPrefix + "/" + basePath;
-    return new RemoteClient(internalAuthenticator, discoveryClient, discoverableServiceName,
-        httpRequestConfig, basePath, remoteAuthenticator);
+      HttpRequestConfig httpRequestConfig, String basePath) {
+    basePath = basePath.startsWith("/") ? pathPrefix + basePath
+        : pathPrefix + "/" + basePath;
+    if (this.useInternalRouter) {
+      return getClientForInternalRouter(discoverableServiceName,
+          httpRequestConfig, basePath);
+    }
+    return new RemoteClient(internalAuthenticator, discoveryClient,
+        discoverableServiceName, httpRequestConfig, basePath,
+        remoteAuthenticator);
+  }
+
+  private RemoteClient getClientForInternalRouter(String destinationServiceName,
+      HttpRequestConfig httpRequestConfig, String basePath) {
+    LOG.trace(
+        "Creating client for service {} which proxies through the {} service."
+        , destinationServiceName, Service.INTERNAL_ROUTER);
+    String proxyPath = String.format("/%s/router/services/%s%s",
+        Constants.Gateway.INTERNAL_API_VERSION_3, destinationServiceName,
+        basePath);
+    return new RemoteClient(internalAuthenticator, discoveryClient,
```

Comment:
There should be another config in the cConf to control the IP of the internal router such that we can use it to construct a static discoveryclient to always use that address?